Make `shouldSuppressTextInputFromEditing` suppress paste from the UI process

charliewolfe opened the pull request:
#### a22b4ce00cb59b781765ba20e28a893471c55053
<pre>
Make `shouldSuppressTextInputFromEditing` suppress paste from the UI process
<a href="https://bugs.webkit.org/show_bug.cgi?id=272345">https://bugs.webkit.org/show_bug.cgi?id=272345</a>
<a href="https://rdar.apple.com/126089620">rdar://126089620</a>

Reviewed by Alex Christensen and Wenson Hsieh.

`shouldSuppressTextInputFromEditing` is used to prevent text input while Safari’s URL bar is showing
a provisional URL. Instead of suppressing the paste event from the web process, this patch makes it so
we do not allow the page to access pasteboard data during a provisional navigation. Paste events will be
sent now, but text on the clipboard will still not be pasted.

This is work towards making paste work with site isolation.

* LayoutTests/http/tests/navigation/keyboard-events-during-provisional-navigation-expected.txt:
* LayoutTests/http/tests/navigation/keyboard-events-during-provisional-navigation.html:
* LayoutTests/http/tests/navigation/keyboard-events-during-provisional-subframe-navigation-expected.txt:
* LayoutTests/http/tests/navigation/keyboard-events-during-provisional-subframe-navigation.html:
* LayoutTests/http/tests/site-isolation/paste-in-iframe-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/paste-in-iframe.html: Added.
* LayoutTests/http/tests/site-isolation/resources/copy-and-paste-text.html: Added.
* LayoutTests/platform/mac-wk2/http/tests/navigation/keyboard-events-during-provisional-navigation-expected.txt:
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::paste):
(WebCore::Editor::pasteAsPlainText):
(WebCore::Editor::pasteAsQuotation):
(WebCore::Editor::pasteFont):
(WebCore::Editor::canPaste const): Deleted.
* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/EditorCommand.cpp:
(WebCore::enabledPaste):
(WebCore::createCommandMap):
(WebCore::allowExecutionWhenDisabledPaste): Deleted.
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::checkOrEnableIfNeeded const):
* Source/WebKit/UIProcess/Cocoa/WebPasteboardProxyCocoa.mm:
(WebKit::WebPasteboardProxy::accessType const):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::editorState const):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView validateUserInterfaceItemWithoutDelegate:]):

Canonical link: <a href="https://commits.webkit.org/277221@main">https://commits.webkit.org/277221@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1c89e85d3656afc845fd5858911b28d94a1b782f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47031 "Failed to checkout and rebase branch from PR 26983") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26200 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49667 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49713 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/43078 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [❌ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49338 "Failed to checkout and rebase branch from PR 26983") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31373 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23666 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47612 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/23658 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/40525 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/19616 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/21026 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/41672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5076 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/43405 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/42066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51590 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/22052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/18420 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/45601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23332 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/40677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/44595 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/24112 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6598 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/23043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on LayoutTests/http/tests/navigation/keyboard-events-during-provisional-subframe-navigation-expected.txt:
```diff
@@ -8,6 +8,8 @@ CONSOLE MESSAGE: Pressing "a".
 CONSOLE MESSAGE: Setting marked text to "b".
 CONSOLE MESSAGE: Inserting text "c".
 CONSOLE MESSAGE: Pasting text "d".
+CONSOLE MESSAGE: pasteevent dispatched (isTrusted: true).
```
Perhaps this could also verify that the data is inaccessible as well? (i.e., verify that `DataTransfer.getData("text/plain")`, `DataTransfer.types` and `DataTransfer.items` all return empty strings/arrays).

charliewolfe replied:
Yeah, I can add this. Maybe line 13 below already tests part of this? Without the change in WebPasteBoardProxy line 13 turns into `CONSOLE MESSAGE: Input element value after text input events: "d".` (d is pasted)

whsieh replied:
Ah, I see! It should be fine as-is, then.

charliewolfe committed:
Make `shouldSuppressTextInputFromEditing` suppress paste from the UI process
https://bugs.webkit.org/show_bug.cgi?id=272345
rdar://126089620

Reviewed by Alex Christensen and Wenson Hsieh.

`shouldSuppressTextInputFromEditing` is used to prevent text input while Safari’s URL bar is showing
a provisional URL. Instead of suppressing the paste event from the web process, this patch makes it so
we do not allow the page to access pasteboard data during a provisional navigation. Paste events will be
sent now, but text on the clipboard will still not be pasted.

This is work towards making paste work with site isolation.

* LayoutTests/http/tests/navigation/keyboard-events-during-provisional-navigation-expected.txt:
* LayoutTests/http/tests/navigation/keyboard-events-during-provisional-navigation.html:
* LayoutTests/http/tests/navigation/keyboard-events-during-provisional-subframe-navigation-expected.txt:
* LayoutTests/http/tests/navigation/keyboard-events-during-provisional-subframe-navigation.html:
* LayoutTests/http/tests/site-isolation/paste-in-iframe-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/paste-in-iframe.html: Added.
* LayoutTests/http/tests/site-isolation/resources/copy-and-paste-text.html: Added.
* LayoutTests/platform/mac-wk2/http/tests/navigation/keyboard-events-during-provisional-navigation-expected.txt:
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::paste):
(WebCore::Editor::pasteAsPlainText):
(WebCore::Editor::pasteAsQuotation):
(WebCore::Editor::pasteFont):
(WebCore::Editor::canPaste const): Deleted.
* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/EditorCommand.cpp:
(WebCore::enabledPaste):
(WebCore::createCommandMap):
(WebCore::allowExecutionWhenDisabledPaste): Deleted.
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::checkOrEnableIfNeeded const):
* Source/WebKit/UIProcess/Cocoa/WebPasteboardProxyCocoa.mm:
(WebKit::WebPasteboardProxy::accessType const):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::editorState const):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView validateUserInterfaceItemWithoutDelegate:]):

Canonical link: https://commits.webkit.org/277221@main
